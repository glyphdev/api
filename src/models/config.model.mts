/**
 * @file Models - Config
 * @module api/config/models/Config
 */

import {
  IsIn,
  IsNotEmpty,
  IsOptional,
  IsString,
  validateSync as validate,
  type ValidationError
} from 'class-validator'

/**
 * Configuration environment.
 *
 * @class
 */
class Config {
  /**
   * The hostname or IP address where the application will listen for incoming
   * connections.
   *
   * @public
   * @instance
   * @member {string} HOST
   */
  @IsString()
  @IsOptional()
  public HOST: string

  /**
   * The type of environment the Node.js process is running in.
   *
   * @public
   * @instance
   * @member {string} NODE_ENV
   */
  @IsIn(['development', 'test', 'production'])
  @IsNotEmpty()
  public NODE_ENV!: string

  /**
   * The port number the application will listen for incoming connections on.
   *
   * @public
   * @instance
   * @member {string} PORT
   */
  @IsString()
  @IsNotEmpty()
  public PORT!: string

  /**
   * Create a new configuration environment.
   *
   * @param {NodeJS.Dict<string>} env
   *  Environment variables object
   */
  constructor(env: NodeJS.Dict<string>) {
    Object.assign(this, env)

    /**
     * Validation errors.
     *
     * @const {ValidationError[]} errors
     */
    const errors: ValidationError[] = validate(this, {
      enableDebugMessages: true,
      forbidUnknownValues: true,
      skipNullProperties: false,
      skipUndefinedProperties: false,
      stopAtFirstError: false,
      validationError: { target: false },
      whitelist: true
    })

    if (errors.length) {
      throw new AggregateError(errors, 'Invalid config')
    }

    this.HOST ||= 'localhost'
  }
}

export default Config
