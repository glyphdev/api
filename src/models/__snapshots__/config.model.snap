// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:models/Config > constructor > should throw if config is invalid 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isIn": "NODE_ENV must be one of the following values: development, test, production",
      "isNotEmpty": "NODE_ENV should not be empty",
    },
    "property": "NODE_ENV",
    "value": undefined,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isNotEmpty": "PORT should not be empty",
      "isString": "PORT must be a string",
    },
    "property": "PORT",
    "value": undefined,
  },
]
`;
